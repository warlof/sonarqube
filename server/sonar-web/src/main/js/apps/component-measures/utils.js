/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import { groupBy, memoize, sortBy, toPairs } from 'lodash';
import { getLocalizedMetricName } from '../../helpers/l10n';
import { cleanQuery, parseAsString, serializeString } from '../../helpers/query';
import { domains } from './config/domains';
import type { Query } from './types';
import type { RawQuery } from '../../helpers/query';
import type { Measure, MeasureEnhanced } from '../../components/measure/types';

export const DEFAULT_VIEW = 'list';
const KNOWN_DOMAINS = [
  'Releasability',
  'Reliability',
  'Security',
  'Maintainability',
  'Coverage',
  'Duplications',
  'Size',
  'Complexity'
];
const BANNED_MEASURES = [
  'blocker_violations',
  'new_blocker_violations',
  'critical_violations',
  'new_critical_violations',
  'major_violations',
  'new_major_violations',
  'minor_violations',
  'new_minor_violations',
  'info_violations',
  'new_info_violations'
];

export function filterMeasures(measures: Array<MeasureEnhanced>): Array<MeasureEnhanced> {
  return measures.filter(measure => !BANNED_MEASURES.includes(measure.metric.key));
}

export function sortMeasures(
  domainName: string,
  measures: Array<MeasureEnhanced>
): Array<MeasureEnhanced> {
  const config = domains[domainName] || {};
  const configOrder = config.order || [];
  return sortBy(measures, [
    measure => {
      const idx = configOrder.indexOf(measure.metric.key);
      return idx >= 0 ? idx : configOrder.length;
    },
    measure => getLocalizedMetricName(measure.metric)
  ]);
}

export function getLeakValue(measure: ?Measure): ?string {
  if (!measure || !measure.periods) {
    return null;
  }
  const period = measure.periods.find(period => period.index === 1);
  return period ? period.value : null;
}

export const groupByDomains = memoize((measures: Array<MeasureEnhanced>): Array<{
  name: string,
  measures: Array<MeasureEnhanced>
}> => {
  const domains = toPairs(groupBy(measures, measure => measure.metric.domain)).map(r => {
    const [name, measures] = r;
    const sortedMeasures = sortBy(measures, measure => getLocalizedMetricName(measure.metric));
    return { name, measures: sortedMeasures };
  });

  return sortBy(domains, [
    domain => {
      const idx = KNOWN_DOMAINS.indexOf(domain.name);
      return idx >= 0 ? idx : KNOWN_DOMAINS.length;
    },
    'name'
  ]);
});

export const parseQuery = memoize((urlQuery: RawQuery): Query => ({
  metric: parseAsString(urlQuery['metric']),
  view: parseAsString(urlQuery['view']) || DEFAULT_VIEW
}));

export const serializeQuery = memoize((query: Query): RawQuery => {
  return cleanQuery({
    metric: serializeString(query.metric),
    view: query.view === DEFAULT_VIEW ? null : serializeString(query.view)
  });
});
