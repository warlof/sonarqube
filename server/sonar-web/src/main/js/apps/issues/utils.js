/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import { searchMembers } from '../../api/organizations';
import { searchUsers } from '../../api/users';
import { formatMeasure } from '../../helpers/measures';
import {
  queriesEqual,
  cleanQuery,
  parseAsBoolean,
  parseAsFacetMode,
  parseAsArray,
  parseAsString,
  serializeString,
  serializeStringArray
} from '../../helpers/query';
import type { RawQuery } from '../../helpers/query';

export type Query = {|
  assigned: boolean,
  assignees: Array<string>,
  authors: Array<string>,
  createdAfter: string,
  createdAt: string,
  createdBefore: string,
  createdInLast: string,
  directories: Array<string>,
  facetMode: string,
  files: Array<string>,
  issues: Array<string>,
  languages: Array<string>,
  modules: Array<string>,
  projects: Array<string>,
  resolved: boolean,
  resolutions: Array<string>,
  rules: Array<string>,
  sort: string,
  severities: Array<string>,
  sinceLeakPeriod: boolean,
  statuses: Array<string>,
  tags: Array<string>,
  types: Array<string>
|};

export type Paging = {
  pageIndex: number,
  pageSize: number,
  total: number
};

// allow sorting by CREATION_DATE only
const parseAsSort = (sort: string): string => (sort === 'CREATION_DATE' ? 'CREATION_DATE' : '');

export const parseQuery = (query: RawQuery): Query => ({
  assigned: parseAsBoolean(query.assigned),
  assignees: parseAsArray(query.assignees, parseAsString),
  authors: parseAsArray(query.authors, parseAsString),
  createdAfter: parseAsString(query.createdAfter),
  createdAt: parseAsString(query.createdAt),
  createdBefore: parseAsString(query.createdBefore),
  createdInLast: parseAsString(query.createdInLast),
  directories: parseAsArray(query.directories, parseAsString),
  facetMode: parseAsFacetMode(query.facetMode),
  files: parseAsArray(query.fileUuids, parseAsString),
  issues: parseAsArray(query.issues, parseAsString),
  languages: parseAsArray(query.languages, parseAsString),
  modules: parseAsArray(query.moduleUuids, parseAsString),
  projects: parseAsArray(query.projectUuids, parseAsString),
  resolved: parseAsBoolean(query.resolved),
  resolutions: parseAsArray(query.resolutions, parseAsString),
  rules: parseAsArray(query.rules, parseAsString),
  sort: parseAsSort(query.s),
  severities: parseAsArray(query.severities, parseAsString),
  sinceLeakPeriod: parseAsBoolean(query.sinceLeakPeriod, false),
  statuses: parseAsArray(query.statuses, parseAsString),
  tags: parseAsArray(query.tags, parseAsString),
  types: parseAsArray(query.types, parseAsString)
});

export const getOpen = (query: RawQuery) => query.open;

export const areMyIssuesSelected = (query: RawQuery): boolean => query.myIssues === 'true';

export const serializeQuery = (query: Query): RawQuery => {
  const filter = {
    assigned: query.assigned ? undefined : 'false',
    assignees: serializeStringArray(query.assignees),
    authors: serializeStringArray(query.authors),
    createdAfter: serializeString(query.createdAfter),
    createdAt: serializeString(query.createdAt),
    createdBefore: serializeString(query.createdBefore),
    createdInLast: serializeString(query.createdInLast),
    directories: serializeStringArray(query.directories),
    facetMode: query.facetMode === 'effort' ? serializeString(query.facetMode) : undefined,
    fileUuids: serializeStringArray(query.files),
    issues: serializeStringArray(query.issues),
    languages: serializeStringArray(query.languages),
    moduleUuids: serializeStringArray(query.modules),
    projectUuids: serializeStringArray(query.projects),
    resolved: query.resolved ? undefined : 'false',
    resolutions: serializeStringArray(query.resolutions),
    s: serializeString(query.sort),
    severities: serializeStringArray(query.severities),
    sinceLeakPeriod: query.sinceLeakPeriod ? 'true' : undefined,
    statuses: serializeStringArray(query.statuses),
    rules: serializeStringArray(query.rules),
    tags: serializeStringArray(query.tags),
    types: serializeStringArray(query.types)
  };
  return cleanQuery(filter);
};

export const areQueriesEqual = (a: RawQuery, b: RawQuery) =>
  queriesEqual(parseQuery(a), parseQuery(b));

type RawFacet = {
  property: string,
  values: Array<{ val: string, count: number }>
};

export type Facet = { [string]: number };

export const mapFacet = (facet: string): string => {
  const propertyMapping = {
    files: 'fileUuids',
    modules: 'moduleUuids',
    projects: 'projectUuids'
  };
  return propertyMapping[facet] || facet;
};

export const parseFacets = (facets: Array<RawFacet>): { [string]: Facet } => {
  // for readability purpose
  const propertyMapping = {
    fileUuids: 'files',
    moduleUuids: 'modules',
    projectUuids: 'projects'
  };

  const result = {};
  facets.forEach(facet => {
    const values = {};
    facet.values.forEach(value => {
      values[value.val] = value.count;
    });
    const finalProperty = propertyMapping[facet.property] || facet.property;
    result[finalProperty] = values;
  });
  return result;
};

export const formatFacetStat = (stat: ?number, mode: string): ?string => {
  if (stat != null) {
    return formatMeasure(stat, mode === 'effort' ? 'SHORT_WORK_DUR' : 'SHORT_INT');
  }
};

export type ReferencedComponent = {
  key: string,
  name: string,
  organization: string,
  path: string
};

export type ReferencedUser = {
  avatar: string,
  name: string
};

export type ReferencedLanguage = {
  name: string
};

export type Component = {
  key: string,
  name: string,
  organization: string,
  qualifier: string
};

export type CurrentUser =
  | { isLoggedIn: false }
  | { isLoggedIn: true, email?: string, login: string, name: string };

export const searchAssignees = (query: string, organization?: string) => {
  return organization
    ? searchMembers({ organization, ps: 50, q: query }).then(response =>
        response.users.map(user => ({
          avatar: user.avatar,
          label: user.name,
          value: user.login
        }))
      )
    : searchUsers(query, 50).then(response =>
        response.users.map(user => ({
          // TODO this WS returns no avatar
          avatar: user.avatar,
          email: user.email,
          label: user.name,
          value: user.login
        }))
      );
};

const LOCALSTORAGE_KEY = 'sonarqube.issues.default';
const LOCALSTORAGE_MY = 'my';
const LOCALSTORAGE_ALL = 'all';

export const isMySet = (): boolean => {
  const setting = window.localStorage.getItem(LOCALSTORAGE_KEY);
  return setting === LOCALSTORAGE_MY;
};

const save = (value: string) => {
  try {
    window.localStorage.setItem(LOCALSTORAGE_KEY, value);
  } catch (e) {
    // usually that means the storage is full
    // just do nothing in this case
  }
};

export const saveMyIssues = (myIssues: boolean) =>
  save(myIssues ? LOCALSTORAGE_MY : LOCALSTORAGE_ALL);
