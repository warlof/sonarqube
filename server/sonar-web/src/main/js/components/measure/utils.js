/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import {
  formatMeasure,
  formatMeasureVariation,
  getRatingTooltip as nextGetRatingTooltip,
  isDiffMetric
} from '../../helpers/measures';
import type { Measure, MeasureEnhanced } from './types';
import type { Metric } from '../../store/metrics/actions';

const KNOWN_RATINGS = ['sqale_rating', 'reliability_rating', 'security_rating'];

export const enhanceMeasure = (measure: Measure, metric: Metric): MeasureEnhanced => ({
  value: measure.value,
  periods: measure.periods,
  metric,
  leak: getLeakValue(measure)
});

export function formatLeak(value: ?string, metric: Metric, options: Object) {
  if (isDiffMetric(metric.key)) {
    return formatMeasure(value, metric.type, options);
  } else {
    return formatMeasureVariation(value, metric.type, options);
  }
}

export function getLeakValue(measure: ?Measure): ?string {
  if (!measure || !measure.periods) {
    return null;
  }
  const period = measure.periods.find(period => period.index === 1);
  return period ? period.value : null;
}

export function getRatingTooltip(metricKey: string, value: ?string) {
  const finalMetricKey = isDiffMetric(metricKey) ? metricKey.substr(4) : metricKey;
  if (KNOWN_RATINGS.includes(finalMetricKey)) {
    return nextGetRatingTooltip(finalMetricKey, value);
  }
  return null;
}
