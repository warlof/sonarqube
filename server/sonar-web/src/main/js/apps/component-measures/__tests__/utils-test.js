/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import * as utils from '../utils';

const MEASURES = [
  {
    metric: {
      key: 'lines_to_cover',
      type: 'INT',
      name: 'Lines to Cover',
      domain: 'Coverage'
    },
    value: '431',
    periods: [{ index: 1, value: '70' }],
    leak: '70'
  },
  {
    metric: {
      key: 'coverage',
      type: 'PERCENT',
      name: 'Coverage',
      domain: 'Coverage'
    },
    value: '99.3',
    periods: [{ index: 1, value: '0.0999999999999943' }],
    leak: '0.0999999999999943'
  },
  {
    metric: {
      key: 'duplicated_lines_density',
      type: 'PERCENT',
      name: 'Duplicated Lines (%)',
      domain: 'Duplications'
    },
    value: '3.2',
    periods: [{ index: 1, value: '0.0' }],
    leak: '0.0'
  }
];

describe('filterMeasures', () => {
  it('should exclude banned measures', () => {
    expect(
      utils.filterMeasures([
        { metric: { key: 'bugs', name: 'Bugs', type: 'INT' } },
        { metric: { key: 'critical_violations', name: 'Critical Violations', type: 'INT' } }
      ])
    ).toHaveLength(1);
  });
});

describe('sortMeasures', () => {
  it('should sort based on the config', () => {
    expect(
      utils.sortMeasures('Reliability', [
        { metric: { key: 'reliability_remediation_effort', name: 'new_bugs', type: 'INT' } },
        { metric: { key: 'new_reliability_remediation_effort', name: 'bugs', type: 'INT' } },
        { metric: { key: 'new_bugs', name: 'new_bugs', type: 'INT' } },
        { metric: { key: 'bugs', name: 'bugs', type: 'INT' } }
      ])
    ).toMatchSnapshot();
  });
});

describe('groupByDomains', () => {
  it('should correctly group by domains', () => {
    expect(utils.groupByDomains(MEASURES)).toMatchSnapshot();
  });

  it('should be memoized', () => {
    expect(utils.groupByDomains(MEASURES)).toBe(utils.groupByDomains(MEASURES));
  });
});

describe('parseQuery', () => {
  it('should correctly parse the url query', () => {
    expect(utils.parseQuery({})).toEqual({ metric: '', selected: '', view: utils.DEFAULT_VIEW });
    expect(utils.parseQuery({ metric: 'foo', selected: 'bar', view: 'tree' })).toEqual({
      metric: 'foo',
      selected: 'bar',
      view: 'tree'
    });
  });

  it('should be memoized', () => {
    const query = { metric: 'foo', selected: 'bar', view: 'tree' };
    expect(utils.parseQuery(query)).toBe(utils.parseQuery(query));
  });
});

describe('serializeQuery', () => {
  it('should correctly serialize the query', () => {
    expect(utils.serializeQuery({ metric: '', selected: '', view: 'list' })).toEqual({});
    expect(utils.serializeQuery({ metric: 'foo', selected: 'bar', view: 'tree' })).toEqual({
      metric: 'foo',
      selected: 'bar',
      view: 'tree'
    });
  });

  it('should be memoized', () => {
    const query = { metric: 'foo', selected: 'bar', view: 'tree' };
    expect(utils.serializeQuery(query)).toBe(utils.serializeQuery(query));
  });
});
