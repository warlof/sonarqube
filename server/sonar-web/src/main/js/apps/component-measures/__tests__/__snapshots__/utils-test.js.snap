// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groupByDomains should correctly group by domains 1`] = `
Array [
  Object {
    "measures": Array [
      Object {
        "leak": "0.0999999999999943",
        "metric": Object {
          "domain": "Coverage",
          "key": "coverage",
          "name": "Coverage",
          "type": "PERCENT",
        },
        "periods": Array [
          Object {
            "index": 1,
            "value": "0.0999999999999943",
          },
        ],
        "value": "99.3",
      },
      Object {
        "leak": "70",
        "metric": Object {
          "domain": "Coverage",
          "key": "lines_to_cover",
          "name": "Lines to Cover",
          "type": "INT",
        },
        "periods": Array [
          Object {
            "index": 1,
            "value": "70",
          },
        ],
        "value": "431",
      },
    ],
    "name": "Coverage",
  },
  Object {
    "measures": Array [
      Object {
        "leak": "0.0",
        "metric": Object {
          "domain": "Duplications",
          "key": "duplicated_lines_density",
          "name": "Duplicated Lines (%)",
          "type": "PERCENT",
        },
        "periods": Array [
          Object {
            "index": 1,
            "value": "0.0",
          },
        ],
        "value": "3.2",
      },
    ],
    "name": "Duplications",
  },
]
`;

exports[`sortMeasures should sort based on the config 1`] = `
Array [
  Object {
    "metric": Object {
      "key": "bugs",
      "name": "bugs",
      "type": "INT",
    },
  },
  Object {
    "metric": Object {
      "key": "new_bugs",
      "name": "new_bugs",
      "type": "INT",
    },
  },
  Object {
    "metric": Object {
      "key": "reliability_remediation_effort",
      "name": "new_bugs",
      "type": "INT",
    },
  },
  Object {
    "metric": Object {
      "key": "new_reliability_remediation_effort",
      "name": "bugs",
      "type": "INT",
    },
  },
]
`;
